lib/server/push.api.js

```diff
@@ -568,88 +568,96 @@ Push.Configure = function(options) {
-                if (notification.query && notification.query === ''+notification.query) {
-                  try {
-                    // The query is in string json format - we need to parse it
-                    notification.query = JSON.parse(notification.query);
-                  } catch(err) {
-                    // Did the user tamper with this??
-                    throw new Error('Push: Error while parsing query string, Error: ' + err.message);
-                  }
-                }
-
-                // Send the notification
-                var result = Push.serverSend(notification);
-
-                if (!options.keepNotifications) {
-                    // Pr. Default we will remove notifications
-                    Push.notifications.remove({ _id: notification._id });
-                } else {
+          try {
+              // Set send fence
+              isSendingNotification = true;
+
+              // var countSent = 0;
+              var batchSize = options.sendBatchSize || 1;
+  
+              // Find notifications that are not being or already sent
+              var pendingNotifications = Push.notifications.find({ $and: [
+                    // Message is not sent
+                    { sent : { $ne: true } },
+                    // And not being sent by other instances
+                    { sending: { $ne: true } },
+                    // And not queued for future
+                    { $or: [ { delayUntil: { $exists: false } }, { delayUntil:  { $lte: new Date() } } ] }
+                ]}, {
+                  // Sort by created date
+                  sort: { createdAt: 1 },
+                  limit: batchSize
+                });
 
-                    // Update the notification
-                    Push.notifications.update({ _id: notification._id }, {
-                        $set: {
-                          // Mark as sent
-                          sent: true,
-                          // Set the sent date
-                          sentAt: new Date(),
-                          // Count
-                          count: result,
-                          // Not being sent anymore
-                          sending: false
-                        }
-                    });
+              pendingNotifications.forEach(function(notification) {
+                  // Reserve notification
+                  var reserved = Push.notifications.update({ $and: [
+                    // Try to reserve the current notification
+                    { _id: notification._id },
+                    // Make sure no other instances have reserved it
+                    { sending: { $ne: true } }
+                  ]}, {
+                    $set: {
+                      // Try to reserve
+                      sending: true
+                    }
+                  });
+
+                  // Make sure we only handle notifications reserved by this
+                  // instance
+                  if (reserved) {
```

Comment:
if reserved and failing we should remove "sending" on the notification - or use a timestamp instead of boolean value.

eg.:
```js
var now = +new Date();
var timeoutAt = now + 60000; // one minute - should be a default option
var reserved = Push.notifications.update({
  { _id: notification._id },
  { sent: false }, // xxx: need to make sure this is set on create
  { sending: { $lt: now } }
},
{
  $set: {
    sending: timeoutAt,
  }
});
```
*also changed to use the implicite "and" instead*